CORPSE_deriv.py

```diff
@@ -86,54 +99,89 @@ def CORPSE_deriv(SOM,T,theta,params,claymod=1.0):
-
-    # Growth is carbon limited and extra N is mineralized
-    loc_Clim=~(loc_Nlim | loc_immob)
-    dmicrobeC[loc_Clim]=(carbon_supply[loc_Clim] - microbeTurnover[loc_Clim]) ;
-    dmicrobeN[loc_Clim]=((carbon_supply[loc_Clim]-maintenance_resp[loc_Clim])/params['CN_microbe'] - microbeTurnover[loc_Clim]*params['et']/params['CN_microbe'])
-    CN_imbalance_term[loc_Clim] = nitrogen_supply[loc_Clim] - (carbon_supply[loc_Clim]-maintenance_resp[loc_Clim])/params['CN_microbe'];
+    microbeTurnover = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    maintenance_resp = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    overflow_resp = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    carbon_supply = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    nitrogen_supply = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    deadmic_C_production = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    deadmic_N_production = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    dmicrobeC = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    dmicrobeN = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    CN_imbalance_term = {'SAP':0.0,'ECM':0.0,'AM':0.0}
+    # Separate microbial metabolisms among different microbial types
+
+    Nacq_simb_max = {'ECM':0.0,'AM':0.0}
+    Nmining = NminingRate(SOM,T,theta,params)
+    Nacq_simb_max['ECM'] = sum(Nmining.values())
+    Nacq_simb_max['AM'] = params['max_scavenging_rate'] * SOM['inorganicN']/(SOM['inorganicN']+params['kc_scavenging_IN']*params['depth']) \
+                     * SOM['AMC']/(SOM['AMC']+params['kc_scavenging']*params['depth'])
+    # Calculate potential ECM N mining and AM N scavenging
+
+    for mt in mic_types:
+        microbeTurnover[mt]=(SOM[mt+'C']-params['minMicrobeC']*(sumCtypes(SOM,'u')))/params['Tmic'][mt];   # kg/m2/yr
+        if isinstance(microbeTurnover[mt],float):
+           microbeTurnover[mt]=max(0.0,microbeTurnover[mt])
+        else:
+           microbeTurnover[mt][microbeTurnover[mt]<0.0]=0.0
+
+        maintenance_resp[mt]=microbeTurnover[mt]*(1.0-et[mt])
+
+        deadmic_C_production[mt]=microbeTurnover[mt]*et[mt]   # actual fraction of microbial turnover
+        deadmic_N_production[mt]=microbeTurnover[mt]*et[mt]/params['CN_microbe'][mt]
+        # Note that we haven't set up mycorrhizal N cycle yet thus we only consider SAP CN interaction for now.
+
+        # C and N available for microbial growth
+        if mt=='SAP':
+           for t in chem_types:
+               carbon_supply[mt]=carbon_supply[mt]+decomp[t+'C']*params['eup'][t]
+               nitrogen_supply[mt]=nitrogen_supply[mt]+decomp[t+'N']*params['nup'][t]
+           IMM_N_max=atleast_1d(params['max_immobilization_rate']*365*SOM['inorganicN']/(SOM['inorganicN']+params['max_immobilization_rate']))
+        else:
+           carbon_supply[mt] += Ctransfer[mt]*params['eup_myc'][mt]
+           nitrogen_supply[mt] += Nacq_simb_max[mt]
+           maintenance_resp[mt] += Ctransfer[mt]*(1-params['eup_myc'][mt])  # MYC Respiration from transforming C from plants to growth
+           IMM_N_max=atleast_1d(0.0)
```

Comment:
This assumes mycorrhizae are not capable of immobilizing nitrogen? It would be more flexible to set a max_immobilization_rate parameter for the mycorrhizae that would allow testing different values or assumptions.
If you want to preserve the ability to run this on an array of points, I would write it like this:
```suggestion
           IMM_N_max=atleast_1d(SOM['inorganicN']*0.0)
```

Reply:
Mycorrhizal fungi are caple of immobilizing nitrogen (right now only AMF in this version, but will let ECM takes up inorgN in the future), the reason I didn't code it as IMM_N_max is that I feel this concept of mycorrhizal N uptake fits better with the variable 'nitrogen_supply': "nitrogen_supply[mt] += Nacq_simb_max[mt]" 
Thus I let N immobilization be the 'specialty' of saprotrophs. But I agree with the 'array of points' approach and will correct that accordingly.